src/Infrastructure/Database/Interceptors/AuditableEntityInterceptor.cs

```diff
@@ -1,65 +1,65 @@
-    }
-
-    public void UpdateEntities(DbContext? context)
-    {
-        if (context == null) return;
-
-        foreach (var entry in context.ChangeTracker.Entries<BaseAuditableModel>())
-        {
-            if (entry.State is EntityState.Added or EntityState.Modified || entry.HasChangedOwnedEntities())
-            {
-                var utcNow = _dateTime.GetUtcNow();
-                if (entry.State == EntityState.Added)
-                {
-                    // entry.Entity.CreatedBy = _user.Id;
-                    entry.Entity.Created = utcNow;
-                } 
-                // entry.Entity.LastModifiedBy = _user.Id;
-                entry.Entity.LastModified = utcNow;
-            }
-        }
-    }
-}
-
-public static class Extensions
-{
-    public static bool HasChangedOwnedEntities(this EntityEntry entry) =>
-        entry.References.Any(r => 
-            r.TargetEntry != null && 
-            r.TargetEntry.Metadata.IsOwned() && 
-            (r.TargetEntry.State == EntityState.Added || r.TargetEntry.State == EntityState.Modified));
-}
+﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using QuestSystem.Application.Common.Interfaces;
+using QuestSystem.Domain.Common;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Diagnostics;
+
+namespace QuestSystem.Infrastructure.Data.Interceptors;
+
+public class AuditableEntityInterceptor : SaveChangesInterceptor
+{
+    private readonly TimeProvider _dateTime;
+
+    public AuditableEntityInterceptor(
+        TimeProvider dateTime)
+    {
+        _dateTime = dateTime;
+    }
+
+    public override InterceptionResult<int> SavingChanges(DbContextEventData eventData, InterceptionResult<int> result)
+    {
+        UpdateEntities(eventData.Context);
+
+        return base.SavingChanges(eventData, result);
+    }
+
+    public override ValueTask<InterceptionResult<int>> SavingChangesAsync(DbContextEventData eventData, InterceptionResult<int> result, CancellationToken cancellationToken = default)
+    {
+        UpdateEntities(eventData.Context);
+
+        return base.SavingChangesAsync(eventData, result, cancellationToken);
+    }
+
+    public void UpdateEntities(DbContext? context)
+    {
+        if (context == null) return;
+
+        foreach (var entry in context.ChangeTracker.Entries<BaseAuditableModel>())
+        {
+            if (entry.State is EntityState.Added or EntityState.Modified || entry.HasChangedOwnedEntities())
+            {
+                var utcNow = _dateTime.GetUtcNow();
+                if (entry.State == EntityState.Added)
+                {
+                    // entry.Entity.CreatedBy = _user.Id;
+                    entry.Entity.Created = utcNow;
+                } 
+                // entry.Entity.LastModifiedBy = _user.Id;
+                entry.Entity.LastModified = utcNow;
+            }
+        }
```

Comment:
Out of the scope of the change but why do we need this interceptor ?

Reply:
This probably changed during the refactoring/renamimg I did, but it was first included when we started creating the quest system back in April/May.